### PR TITLE
tasks query bug fix

### LINKED ISSUE #3216: schema broken (GraphQLLocatedError for 'state')
Attempting some development today I found that the end-to-end system has broken so that using the current ``master`` on all three repositories ``cylc-flow``, ``cylc-ui`` & ``cylc-uiserver`` means that (at least for the several suites I tested on), the UI Tree View gives a blank page apart from the top bar with the suite name (see the screenshot below), & further digging uncovers some error messages.

I have traced this via reverting commits (on each of the three repos independently until I hit the case where the full system went from working as previous to hitting the above problem) to #3211, which unfortunately has broken the schema, throwing up errors relating to 'state', as below.

I'm going to trace back the logic to work out which aspect(s) of that PR (with one commit) are to blame, & put in a PR to fix them. Thanks.

#### GraphiQL

From interactive GraphQL, I got the following error response returned:

```graphql
  "errors": [
    {
      "message": "state",
      "locations": [
        {
          "line": 21,
          "column": 5
        }
      ],
      "path": [
        "workflows",
        0,
        "tasks"
      ]
    }
  ],
```
(if relevant, this was from a basic query on one of my running workflows:)
```graphql
query workflows($tIds: [ID]){
  workflows(ids: "subsuite-app-domingo"){
    id
    name
    status
    stateTotals {
      runahead
      waiting
      held
      queued
      expired
      ready
      submitFailed
      submitRetrying
      submitted
      retrying
      running
      failed
      succeeded
    }
    tasks(ids: $tIds){
      name
      meta {
        title
        description
        URL
        userDefined
      }
      proxies(ids: $tIds){
        cyclePoint
        state
      }
    }
  }
}
```

### Terminal

And a traceback to the terminal was consistent with it being an issue on ``state``, emerging from the ``schema.py``.
```
ERROR:graphql.execution.utils:Traceback (most recent call last):
  File "/home/h06/sbarth/miniconda3/envs/py3env/lib/python3.7/site-packages/promise/promise.py", line 842, in handle_future_result
    resolve(future.result())
  File "/net/home/h06/sbarth/cylc.git/cylc/flow/network/schema.py", line 263, in get_nodes_by_id
    return await resolvers.get_nodes_by_id(node_type, args)
  File "/net/home/h06/sbarth/cylc-uiserver/resolvers.py", line 65, in get_nodes_by_id
    return [node for node in nodes
  File "/net/home/h06/sbarth/cylc-uiserver/resolvers.py", line 67, in <listcomp>
    node_filter(node, args))]
  File "/net/home/h06/sbarth/cylc.git/cylc/flow/network/resolvers.py", line 90, in node_filter
    (args.get('ghosts') or node.state != '') and
graphql.error.located_error.GraphQLLocatedError: state
```

**************************

**Release version(s) and/or repository branch(es) affected?**
``8.0a0``.

**Steps to reproduce the bug**

* Update all three repos to the current ``master`` branches.
* Start up the UI Server & open the UI, as usual.
* Run some suites.
* Open them in Tree View, & see that the page is there but the table is missing, with no data there emerging.

**Expected behavior**

* A Tree View (nested) table populated with its task data, for each suite.

**Screenshots**

| Current ``master`` | After reverting commit from #3211 |
|---|---|
| Screenshot from 2019-07-13 19-02-04  | Screenshot from 2019-07-13 19-06-10 |

### PR DESCRIPTION
**Pull request by dwsutherland**
These changes close #3216

Test with bash script:
```
#!/bin/bash

# wsgraphql <workflow_name> <request_args_name>

read -r -d '' flatTree <<_args_
{
  "request_string": "
query flatTree(\$wIds: [ID], \$tIds: [ID], \$nIds: [ID], \$nStates: [String], \$minDepth: Int, \$maxDepth: Int) {
  workflows(ids: \$wIds) {
    id
    name
    status
    stateTotals {
      runahead
      waiting
      held
      queued
      expired
      ready
      submitFailed
      submitRetrying
      submitted
      retrying
      running
      failed
      succeeded
    }
    treeDepth
    tasks(ids: \$tIds, mindepth: \$minDepth, maxdepth: \$maxDepth){
      name
      meta {
        title
        description
        URL
        userDefined
      }
      proxies(ids: \$nIds, states: \$nStates){
        cyclePoint
        state
      }
    }
  }
}",
"variables": {
    "wIds": ["*/${1}"],
    "nIds": ["20*/*"],
    "tIds": ["*"],
    "nStates": ["succeeded", "waiting", "held", "running"],
    "minDepth": 0,
    "maxDepth": 4
  }
}
_args_

# Invoke command with given request
echo "${!2}"
echo ${!2} | time cylc client $1 graphql
```
```
$ wsgraphql baz flatTree
```

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] This is a bug fix to existing code with tests pending further development.
- [x] No change log entry required (e.g. change is small or internal only).
- [x] No documentation update required for this change.

### CODE REVIEW

**Comment by kinow** [comment]
Oh, so you actually didn't need to add state to the task! Thanks for the fix so quickly, but it could wait until you were back :+1:

**Comment by dwsutherland** [comment]
> Oh, so you actually didn't need to add state to the task! Thanks for the fix so quickly, but it could wait until you were back 👍

Roger that! back to Holiday! (just thought this was blocking progress so wanted to respond)

**Comment by kinow** [comment]
Don't worry, it is definitely not blocking. We don't have any new features, so at any time we can simply git checkout a commit before the one where I broke master :) enjoy holidays!!!

File: cylc/flow/network/resolvers.py [hunk 1]
```diff
@@ -87,10 +87,10 @@ def node_filter(node, args):
     """Filter nodes based on attribute arguments"""
     natts = collate_node_atts(node)
     return (
-        (args.get('ghosts') or node.state != '') and
-        (not (args.get('states') and node.state != '')
+        (args.get('ghosts') or natts[5] != '') and
+        (not (args.get('states') and natts[5] != '')
          or node.state in args['states']) and
-        not (args.get('exstates') and node.state != '' and
+        not (args.get('exstates') and natts[5] != '' and
```
Comment (matthewrmshin) [inline cylc/flow/network/resolvers.py]: Maybe add a comment above to explain what the elements in `natts` are?

**Comment by matthewrmshin** [comment]
Worth adding an automated test? Or defer?

**Comment by kinow** [comment]
>Worth adding an automated test? Or defer?

I tried writing a unit test, and it was taking a bit longer than I expected (>1hour), and I was starting to modify parts of the code to make it easier to write tests... then I gave up.

An automated test similar to our functional tests in cylc-flow may be easier, though later after the code base has stabilized more, I guess a small refactor + unit tests will still be necessary.

**Comment by kinow** [comment]
Comment from yesterday in the ticket about the wrong wheel

>The wheel has been deleted. Sorry for any inconvenience this may have caused.

Kicked Travis, and now the unit tests are all passing :hand: 

However, now the functional tests are refusing to pass. Kicked them three times, each time a different output (once it was a failure to produce the coverage reports due to out of memory error :confused: )

**Comment by kinow** [comment]
Functional tests are refusing to pass. Kicked them three times, each time a different output (once it was a failure to produce the coverage reports due to out of memory error :confused: ). Last kick of the day :+1: 4th time will be the charm now :crossed_fingers:

File: cylc/flow/network/resolvers.py (hunk 1 above)
Comment (dwsutherland) [reply cylc/flow/network/resolvers.py]: Done. Hope I it's not too verbose (or not verbose enough).

**Comment by dwsutherland** [comment]
> Worth adding an automated test? Or defer?

Was going to include tests in #3202   ... Could be it's own PR I suppose


> An automated test similar to our functional tests in cylc-flow may be easier, though later after the code base has stabilized more, I guess a small refactor + unit tests will still be necessary.

Yes, the code is under a bit of flux at the moment (especially with #3202 )..
I have a few ideas around functional tests, i.e. 
- Static protobuf message (saved to disk) to interrogate with GraphQL queries (diff may require order preservation, and live data will have date and checksum differences).
- Live suite GraphQL functionality test (set of mutations and queries to cover all).
- Others?

These tests could be used (imported?) by the UIS.

**Comment by kinow** [comment]
Ha! Now the build passed. @matthewrmshin another curious case. Kicked 4 times a job that was failing with weird messages like "Cannot find contact file". Then kicked it again next day, and nothing. Finally, restarted the whole build, and after ~1 hour, it passed.

**Comment by dwsutherland** [comment]
> Ha! Now the build passed. @matthewrmshin another curious case. Kicked 4 times a job that was failing with weird messages like "Cannot find contact file". Then kicked it again next day, and nothing. Finally, restarted the whole build, and after ~1 hour, it passed.

@matthewrmshin - Is it possible that multiple tests (even across chunks) are using the same suites? hence the contact error and inconsistent errors?

**Pull request merged**